server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java

```diff
@@ -1539,6 +1557,126 @@ public RepositoriesService getRepositoriesService() {
         void onSnapshotFailure(Snapshot snapshot, Exception e);
     }
 
+    public static final class SnapshotsInProgressUpdater extends RoutingChangesObserver.AbstractChangedShardObserver {
+
+        private final Set<ShardId> shardChanges = new HashSet<>();
+
+        public SnapshotsInProgress applyChanges(SnapshotsInProgress oldSnapshot, RoutingTable newRoutingTable) {
+            return updateWithRoutingTable(shardChanges, oldSnapshot, newRoutingTable);
+        }
+
+        @Override
+        protected void onChanged(ShardId shardId) {
+            shardChanges.add(shardId);
+        }
+    }
+
+    private static SnapshotsInProgress updateWithRoutingTable(Set<ShardId> shardIds, SnapshotsInProgress oldSnapshot,
+        RoutingTable newRoutingTable) {
+        if (oldSnapshot == null || shardIds.isEmpty()) {
+            return oldSnapshot;
+        }
+        List<SnapshotsInProgress.Entry> entries = new ArrayList<>();
+        boolean snapshotsInProgressChanged = false;
+        for (SnapshotsInProgress.Entry entry : oldSnapshot.entries()) {
+            ImmutableOpenMap.Builder<ShardId, ShardSnapshotStatus> shardsBuilder = null;
+            for (ShardId shardId : shardIds) {
+                ImmutableOpenMap<ShardId, ShardSnapshotStatus> shards = entry.shards();
+                ShardSnapshotStatus currentStatus = shards.get(shardId);
+                if (currentStatus != null && currentStatus.state().completed() == false) {
+                    final ShardSnapshotStatus newStatus = Optional
+                        .ofNullable(newRoutingTable.shardRoutingTableOrNull(shardId))
```

Comment:
I wonder if we can expect there to always exist the corresponding shard routing table?

Reply:
I don't know but can't a `ShardId` that we got from the callback `org.elasticsearch.cluster.routing.RoutingChangesObserver.AbstractChangedShardObserver#shardFailed` lead to a `null` entry here? (I could be way off here ... not sure about the order of these things yet)

Reply:
we are never adding or removing keys in the shard routing table during the `AllocationService.reroute()` phase. Can you add an assertion that the routing table is not null?